src/clients/c++/examples/simple_cuda_shm_client.cc

```diff
@@ -0,0 +1,398 @@
+  if (!err.IsOk()) {
+    std::cerr << "failed setting shared memory input: " << err << std::endl;
+    exit(1);
+  }
+
+  // Send inference request to the inference server.
+  std::map<std::string, std::unique_ptr<nic::InferContext::Result>> results;
+  FAIL_IF_ERR(infer_ctx->Run(&results), "unable to run model");
+
+  // We expect there to be 2 results. Walk over all 16 result elements
+  // and print the sum and difference calculated by the model.
+  if (results.size() != 2) {
+    std::cerr << "error: expected 2 results, got " << results.size()
+              << std::endl;
+  }
+
+  // Copy input and output data back to the CPU
+  int input0_data[16], input1_data[16];
+  int output0_data[16], output1_data[16];
+  cudaMemcpy(
+      input0_data, input0_d_ptr, input_byte_size, cudaMemcpyDeviceToHost);
+  cudaMemcpy(
+      input1_data, input1_d_ptr, input_byte_size, cudaMemcpyDeviceToHost);
+  cudaMemcpy(
+      output0_data, output0_d_ptr, output_byte_size, cudaMemcpyDeviceToHost);
+  cudaMemcpy(
+      output1_data, output1_d_ptr, output_byte_size, cudaMemcpyDeviceToHost);
+
+  for (size_t i = 0; i < 16; ++i) {
+    std::cout << input0_data[i] << " + " << input1_data[i] << " = "
+              << output0_data[i] << std::endl;
+    std::cout << input0_data[i] << " - " << input1_data[i] << " = "
+              << output1_data[i] << std::endl;
+
+    if ((input0_data[i] + input1_data[i]) != output0_data[i]) {
+      std::cerr << "error: incorrect sum" << std::endl;
+      exit(1);
+    }
+    if ((input0_data[i] - input1_data[i]) != output1_data[i]) {
+      std::cerr << "error: incorrect difference" << std::endl;
+      exit(1);
+    }
+  }
+
+  // Get shared memory regions all active/registered within TRTIS
+  ni::SharedMemoryStatus status;
+  err = shared_memory_ctx->GetSharedMemoryStatus(&status);
+  if (!err.IsOk()) {
+    std::cerr << "error: " << err << std::endl;
+    exit(1);
+  }
+  std::cout << "Shared Memory Status:\n" << status.DebugString() << "\n";
+
+  // Unregister shared memory (One by one or all at a time) from TRTIS
+  // err = shared_memory_ctx->UnregisterAllSharedMemory();
+  err = shared_memory_ctx->UnregisterSharedMemory("input_data");
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to unregister shared memory input region: "
+              << err << std::endl;
+    exit(1);
+  }
+  err = shared_memory_ctx->UnregisterSharedMemory("output_data");
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to unregister shared memory output region: "
+              << err << std::endl;
+    exit(1);
+  }
+
+  // Free GPU memory
+  cudaError_t cuerr = cudaFree(input0_d_ptr);
+  if (cuerr != cudaSuccess) {
+    std::cerr << "CUDA free (input0_d_ptr) error: " << cudaGetErrorName(cuerr)
+              << "\n";
+  }
+  cuerr = cudaFree(output0_d_ptr);
+  if (cuerr != cudaSuccess) {
+    std::cerr << "CUDA free (output0_d_ptr) error: " << cudaGetErrorName(cuerr)
+              << "\n";
+  }
```

Comment:
CUDA error handling can be wrapped with the new CUDA macro.

Reply:
Done